verilog/analysis/checkers/one_module_per_file_rule_test.cc

```diff
@@ -0,0 +1,74 @@
+// Copyright 2017-2020 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "verilog/analysis/checkers/one_module_per_file_rule.h"
+
+#include <initializer_list>
+#include <string>
+
+#include "gtest/gtest.h"
+#include "common/analysis/linter_test_utils.h"
+#include "common/analysis/text_structure_linter_test_utils.h"
+#include "common/text/symbol.h"
+#include "verilog/analysis/verilog_analyzer.h"
+#include "verilog/parser/verilog_token_enum.h"
+
+namespace verilog {
+namespace analysis {
+namespace {
+
+using verible::LintTestCase;
+using verible::RunLintTestCases;
+
+// Expected token type of findings.
+constexpr int kToken = SymbolIdentifier;
+
+// Test that no violations are found without module declarations
+TEST(OneModulePerFileRuleTest, NoModules) {
+  const std::initializer_list<LintTestCase> kTestCases = {
+      {""},
+      {"class c; endclass"},
+      {"package q; endpackage"},
+  };
+  RunLintTestCases<VerilogAnalyzer, OneModulePerFileRule>(kTestCases, "");
+}
+
+// Test that sole module in file is not reported
+TEST(OneModulePerFileRuleTest, OneModule) {
+  const std::initializer_list<LintTestCase> kTestCases = {
+      {"module moo; endmodule"},
+      {"class c; endclass\nmodule moo; endmodule"},
+      {"package q; endpackage\nmodule moo; endmodule"},
+  };
+  RunLintTestCases<VerilogAnalyzer, OneModulePerFileRule>(kTestCases, "");
+}
+
+// Test that multiple module declarations are picked up
+TEST(OneModulePerFileRuleTest, MultipleModules) {
+  const std::initializer_list<LintTestCase> kTestCases = {
+      {"module foo; endmodule\n"
+       "module ", {kToken, "moo"}, "; endmodule"},
+      {"class c; endclass\n"
+       "module foo; endmodule\n"
+       "module ", {kToken, "moo"}, "; endmodule"},
+      {"package q; endpackage\n"
+       "module foo; endmodule\n"
+       "module ", {kToken, "moo"}, "; endmodule\n"},
+  };
+  RunLintTestCases<VerilogAnalyzer, OneModulePerFileRule>(kTestCases, "");
+}
```

Comment:
According to the LRM, it looks like module_declaration is allowed as a 'non_port_module_item', which suggests that modules may be nested.  I think we may want to restrict the findings to only outermost module declarations.
Suggested test case:

```systemverilog
module outer;
module inner;
endmodule : inner
endmodule : outer
```

The `module-filename` rule probably needs to be updated to account for this as well, but that can be a separate PR if you like.

Reply:
Good find, thanks! I've adapted both rules to account for that.